contracts/specs/NFTSwapper/NFTSwapperSpec.md

```diff
@@ -0,0 +1,163 @@
+# NFTSwapper Spec
+
+NFTSwapper enables permissionless swapping between ERC721's within a collection. 
+
+This contract is immutable with no owner. No NFTs or ETH are ever held by this contract, instead it just facilitates the permissionless swap process. 
+
+---
+
+- [NFTSwapper Spec](#nftswapper-spec)
+  - [Function Interface](#function-interface)
+    - [`constructor`](#constructor)
+    - [`swapPrice`](#swapprice)
+    - [`swap`](#swap)
+  - [Swap Price Curve](#swap-price-curve)
+    - [Constant Price Curve](#constant-price-curve)
+    - [Output NFT age based curve](#output-nft-age-based-curve)
+    - [Input and Output NFT age difference based curve](#input-and-output-nft-age-difference-based-curve)
+  - [Approval Considerations](#approval-considerations)
+
+## Function Interface
+
+### `constructor`
+
+```solidity
+constructor(
+    address swapPool, 
+    address nft, 
+    address feeRecipient, 
+    PriceCurveParams priceCurveParams
+)
+```
+
+* `swapPool`: the address which holds the swappable nft's, note that the `swapPool` must approve this contract for it's NFT's (or a subset of them, see [Approval Considerations](#approval-considerations)).
+* `nft`: the address of the ERC721 token that is being swapped.
+* `feeRecipient`: the address where swap fees will be sent. 
+* `priceCurveParams`: the parameters for the price curve (see [Swap Price Curve](#swap-price-curve)).
+ 
+### `swapPrice`
+
+```solidity
+function swapPrice(
+    uint256 inputTokenId, 
+    uint256 outputTokenId
+) pure returns (uint256)
+```
+
+Pure function to get the price in Wei to perform a swap from `inputTokenId` to `outputTokenId` based on [swap price curve](#swap-price-curve).
+
+### `swap`
+
+```solidity
+function swap(uint256 inputTokenId, uint256 outputTokenId) payable 
+```
+
+Swaps the `inputTokenId` from `msg.sender` with the `outputTokenId` from the `swapPool`. Sends the swap fee to the `feeRecipient`.  
+
+Requirements:
+* This contract must be approved to transfer the `nft` with `inputTokenId` on behalf of `msg.sender`.
+* This contract must be approved to transfer the `nft` with `outputTokenId` on behalf of the `swapPool`
+* Swap fee of `swapPrice(inputTokenId, outputTokenId)` Wei must be provided 
+
+
+## Swap Price Curve 
+
+The swap price is determined by the curve defined below. The `PriceCurveParams` are configured on deployment and cannot be changed.
+
+> Currently 3 different options are being considered, only 1 will be selected once finalized
+> - [Constant Price Curve](#constant-price-curve)
+> - [Output NFT age based curve](#output-nft-age-based-curve)
+> - [Input and Output NFT age difference based curve](#input-and-output-nft-age-difference-based-curve)
```

Comment:
I'm a fan of constant or input/output. I think output is just a worse version of input/output -- if we are valuing "old" ids then it makes sense that you get a discount if you contribute an older id. But then here's a question: if you swap an older id for a newer id, should you _make_ money? Should it be free? I feel like there are some weirder dynamics here. I don't really like the idea of paying users to swap low for high... then theres an arb where you could potentially buy and swap low for high Ids. It feels like complexity that doesn't need to exist imo.

Reply:
I lean towards "Input and Output NFT age difference", which best preserve fungibility IMO. 

See the price curve image below, the minimum swap fee is still the `floor`, so int he case of older (lower) ID for a newer (higher) ID, they pay the floor. 

Don't think we should pay them ever for a swap, the DAO is still providing a swapping service here, they just get it for the lowest price.

Reply:
Here is the plot I am referring to:

![image](https://github.com/papercliplabs/noun-swap/assets/31705429/31d14e44-f8a1-44c8-a1a0-dfe7d49c12f0)